ibis/common/tests/test_collections.py

```diff
@@ -0,0 +1,119 @@
+from collections.abc import ItemsView, Iterator, KeysView, ValuesView
+
+import pytest
+
+from ibis.common.collections import MapSet
+
+
+class MySchema(MapSet):
+    def __init__(self, dct=None, **kwargs):
+        self._fields = dict(dct or kwargs)
+
+    def __repr__(self):
+        return f'{self.__class__.__name__}({self._fields})'
+
+    def __getitem__(self, key):
+        return self._fields[key]
+
+    def __iter__(self):
+        return iter(self._fields)
+
+    def __len__(self):
+        return len(self._fields)
+
+
+def test_mapset_mapping_api():
+    ms = MySchema(a=1, b=2)
+    assert ms['a'] == 1
+    assert ms['b'] == 2
+    assert len(ms) == 2
+    assert isinstance(iter(ms), Iterator)
+    assert list(ms) == ['a', 'b']
+    assert isinstance(ms.keys(), KeysView)
+    assert list(ms.keys()) == ['a', 'b']
+    assert isinstance(ms.values(), ValuesView)
+    assert list(ms.values()) == [1, 2]
+    assert isinstance(ms.items(), ItemsView)
+    assert list(ms.items()) == [('a', 1), ('b', 2)]
+    assert ms.get('a') == 1
+    assert ms.get('c') is None
+    assert ms.get('c', 3) == 3
+    assert 'a' in ms
+    assert 'c' not in ms
+    assert ms == ms
+    assert ms != MySchema(a=1, b=2, c=3)
+
+
+def test_mapset_set_api():
+    a = MySchema(a=1, b=2)
+    a_ = MySchema(a=1, b=-2)
+    b = MySchema(a=1, b=2, c=3)
+    b_ = MySchema(a=1, b=2, c=-3)
+    c = MySchema(a=1)
+    f = MySchema(d=4, e=5)
+
+    # disjoint
+    assert not a.isdisjoint(b)
+    assert a.isdisjoint(f)
+
+    # __eq__, __ne__
+    assert a == a
+    assert a != a_
+    assert b == b
+    assert b != b_
+
+    # __le__, __lt__
+    assert a < b
+    assert a <= b
+    assert a <= a
+    assert not b <= a
+    assert not b < a
+    assert not a < a_
+    assert not a <= a_
+
+    # __gt__, __ge__
+    assert b > a
+    assert b >= a
+    assert a >= a
+    assert not a >= b
+    assert not a > b
+    assert not a_ > a
+    assert not a_ >= a
+
+    # __and__
+    assert a & a_ == c
+    assert a & b == a
+    assert b & b_ == a
+    assert a & f == MySchema()
+
+    # __or__
+    assert a | a == a
+    assert a | b == b
+    assert a | f == MySchema(a=1, b=2, d=4, e=5)
+    with pytest.raises(ValueError):
+        # duplicate keys with different values
+        a | a_
+
+    # merge (alias for __or__)
+    assert a.merge(a) == a
+    assert a.merge(b) == b
+    with pytest.raises(ValueError):
+        # duplicate keys with different values
+        a.merge(a_)
+
+    # __sub__
+    assert a - a_ == MySchema(b=2)
+    assert a - b == MySchema()
+    assert b - a == MySchema(c=3)
+    assert a - f == a
+    assert f - a == f
+
+    # __xor__
+    with pytest.raises(ValueError):
+        # duplicate keys with different values
+        a ^ a_
+
+    assert a ^ b == MySchema(c=3)
+    assert b ^ a == MySchema(c=3)
+    assert a ^ f == MySchema(a=1, b=2, d=4, e=5)
+    assert f ^ a == MySchema(a=1, b=2, d=4, e=5)
```

Comment:
Why wouldn't `f ^ a` return `MySchema(d=4, e=5, a=1, b=2)`? I would expect the order to be preserved in binary operations.

Reply:
Need better assertions since `__eq__` doesn't check for order.

Reply:
Fixed.